lightning-block-sync/src/http.rs

```diff
@@ -0,0 +1,740 @@
+	host: String,
+	port: Option<u16>,
+	path: String,
+}
+
+impl HttpEndpoint {
+	/// Creates an endpoint for the given host and default HTTP port.
+	pub fn for_host(host: String) -> Self {
+		Self {
+			host,
+			port: None,
+			path: String::from("/"),
+		}
+	}
+
+	/// Specifies a port to use with the endpoint.
+	pub fn with_port(mut self, port: u16) -> Self {
+		self.port = Some(port);
+		self
+	}
+
+	/// Specifies a path to use with the endpoint.
+	pub fn with_path(mut self, path: String) -> Self {
+		self.path = path;
+		self
+	}
+
+	/// Returns the endpoint host.
+	pub fn host(&self) -> &str {
+		&self.host
+	}
+
+	/// Returns the endpoint port.
+	pub fn port(&self) -> u16 {
+		match self.port {
+			None => 80,
+			Some(port) => port,
+		}
+	}
+
+	/// Returns the endpoint path.
+	pub fn path(&self) -> &str {
+		&self.path
+	}
+}
+
+impl<'a> std::net::ToSocketAddrs for &'a HttpEndpoint {
+	type Iter = <(&'a str, u16) as std::net::ToSocketAddrs>::Iter;
+
+	fn to_socket_addrs(&self) -> std::io::Result<Self::Iter> {
+		(self.host(), self.port()).to_socket_addrs()
+	}
+}
+
+/// Client for making HTTP requests.
+pub(crate) struct HttpClient {
+	stream: TcpStream,
+}
+
+impl HttpClient {
+	/// Opens a connection to an HTTP endpoint.
+	pub fn connect<E: ToSocketAddrs>(endpoint: E) -> std::io::Result<Self> {
+		let address = match endpoint.to_socket_addrs()?.next() {
+			None => {
+				return Err(std::io::Error::new(std::io::ErrorKind::InvalidInput, "could not resolve to any addresses"));
+			},
+			Some(address) => address,
+		};
+		let stream = std::net::TcpStream::connect_timeout(&address, Duration::from_secs(1))?;
+		stream.set_read_timeout(Some(Duration::from_secs(2)))?;
+		stream.set_write_timeout(Some(Duration::from_secs(1)))?;
+
+		#[cfg(feature = "tokio")]
+		let stream = {
+			stream.set_nonblocking(true)?;
+			TcpStream::from_std(stream)?
+		};
+
+		Ok(Self { stream })
+	}
+
+	/// Sends a `GET` request for a resource identified by `uri` at the `host`.
+	///
+	/// Returns the response body in `F` format.
+	#[allow(dead_code)]
+	pub async fn get<F>(&mut self, uri: &str, host: &str) -> std::io::Result<F>
+	where F: TryFrom<Vec<u8>, Error = std::io::Error> {
+		let request = format!(
+			"GET {} HTTP/1.1\r\n\
+			 Host: {}\r\n\
+			 Connection: keep-alive\r\n\
+			 \r\n", uri, host);
+		let response_body = self.send_request_with_retry(&request).await?;
+		F::try_from(response_body)
+	}
+
+	/// Sends a `POST` request for a resource identified by `uri` at the `host` using the given HTTP
+	/// authentication credentials.
+	///
+	/// The request body consists of the provided JSON `content`. Returns the response body in `F`
+	/// format.
+	#[allow(dead_code)]
+	pub async fn post<F>(&mut self, uri: &str, host: &str, auth: &str, content: serde_json::Value) -> std::io::Result<F>
+	where F: TryFrom<Vec<u8>, Error = std::io::Error> {
+		let content = content.to_string();
+		let request = format!(
+			"POST {} HTTP/1.1\r\n\
+			 Host: {}\r\n\
```

Comment:
Does this add a bunch of spacing before each header on the wire?

Reply:
Surprisingly, no. Really cool that Rust will remove the leading whitespace when the previous line ends in a backslash:

https://doc.rust-lang.org/reference/tokens.html#string-literals